src/java.base/share/classes/java/net/Socket.java

```diff
@@ -695,15 +703,9 @@ public void connect(SocketAddress endpoint, int timeout) throws IOException {
 
         try {
             getImpl().connect(epoint, timeout);
-        } catch (SocketTimeoutException e) {
-            throw e;
-        } catch (InterruptedIOException e) {
-            Thread thread = Thread.currentThread();
-            if (thread.isVirtual() && thread.isInterrupted()) {
-                close();
-                throw new SocketException("Closed by interrupt");
-            }
-            throw e;
+        } catch (IOException error) {
```

Comment:
At L680 we reject an endpoint that is not an InetSocketAddresss. I think we should follow this with a check to ensure that the address is resolved, that way Socket does all the validation before delegation to the SocketImpl. It means if getImpl().connect(..) throws any exception or error then you can close the Socket, meaning you can catch Throwable.

Reply:
The spec is still not right: we throw SocketException if the socket is already connected, but we do not close the socket in that case.

Reply:
By comparison SocketChannel throws `java.nio.channels.AlreadyConnectedException` which is a subclass of IllegalSstateException (a runtime exception) and leaves the socket open. I am not sure we could change `Socket::connect` to throw a different exception than `SocketException` in that case due to compatibility reasons. So we now have to choose between either:

- modify the spec to explicitely say that the socket remains open if a SocketException is thrown because the socket is already connected
- modify the behavior to also close the socket in that case

Whatever alternative we choose should be also listed in the CSR.

Reply:
@AlanBateman, how shall we proceed with this one?